storage: Carve chunks of maxBytes / 2 off a range when splitting instead of splitting in the middle.

d4l3k opened the pull request:
This minimizes the amount of work done when splitting huge ranges as only ~32MB
needs to have stats recomputed instead of half of a potentially >1GB range.

Fixes #6400.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7118)

<!-- Reviewable:end -->

d4l3k committed:
storage: Carve chunks of maxBytes / 2 off a range when splitting instead of splitting in the middle.

This minimizes the amount of work done when splitting huge ranges as only ~32MB
needs to have stats recomputed instead of half of a potentially >1GB range.

This also removes an old debug command to manually run MVCCFindSplitKey as it's
no longer relevant.

Fixes #6400.

tschottdorf commented:
Could you add commentary to `(*Replica).splitTrigger` that highlights that only the left half is recomputed, and why that's important? Just to make sure that this doesn't accidentally change.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

@cuongdo in line with the tests you're running right now, one that starts a single-node cluster with a wildly oversized range would be useful. This is easy to obtain by changing the zone config before creating the data dump, and then changing it back as you run the test.

<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> #### storage: Carve chunks of maxBytes / 2 off a range when splitting instead of splitting in the middle\.
> 
> This minimizes the amount of work done when splitting huge ranges as only ~32MB
> 
> needs to have stats recomputed instead of half of a potentially >1GB range.
> 
> 
> 
> Fixes #6400.
> 
> 
> 
</details>


---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[storage/engine/mvcc_test.go, line 2745 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7118#-KJloXGVxnY3ZPlHuTuL:-KJloXGWuq7ENPW_cDTq:-2041185795) ([raw file](https://github.com/cockroachdb/cockroach/blob/baab0ac3cd5dd770f1bb43420be104e8be670487/storage/engine/mvcc_test.go#L2745)):*
> ```Go
> 		snap := engine.NewSnapshot()
> 		defer snap.Close()
> 		targetSize := (ms.KeyBytes + ms.ValBytes) / 2
> ```

Could you do a sanity check that a targetSize of `0` and `\infinity` behaves accordingly?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7118#-:-KJlnxw6CePn774ulc2W:-720568726)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
I updated the comment and added a TODO about automatically determining which range is smaller.

<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Could you add commentary to `(*Replica).splitTrigger` that highlights that only the left half is recomputed, and why that's important? Just to make sure that this doesn't accidentally change.
> 
> :lgtm:
> 
> @cuongdo in line with the tests you're running right now, one that starts a single-node cluster with a wildly oversized range would be useful. This is easy to obtain by changing the zone config before creating the data dump, and then changing it back as you run the test.
</details>


---

Review status: 1 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/engine/mvcc_test.go, line 2745 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7118#-KJloXGVxnY3ZPlHuTuL:-KJlwI6CH_v2R87miboo:-1900021004) ([raw file](https://github.com/cockroachdb/cockroach/blob/baab0ac3cd5dd770f1bb43420be104e8be670487/storage/engine/mvcc_test.go#L2745)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Could you do a sanity check that a targetSize of `0` and `\infinity` behaves accordingly?
</details>
Added test cases.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7118#-:-KJlwNvrat6NyJPHdPuz:796334697)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 3 of 3 files at r2.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7118)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Yes, that seems like a useful test. Can you submit an issue for that?

On Wed, Jun 8, 2016 at 5:04 PM Tobias Schottdorf <notifications@github.com>
wrote:

> Could you add commentary to (*Replica).splitTrigger that highlights that
> only the left half is recomputed, and why that's important? Just to make
> sure that this doesn't accidentally change.
>
> [image: <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>]
> <https://camo.githubusercontent.com/41b3ec3419116e3b3f84507ad965646ce4ce1f0c/68747470733a2f2f72657669657761626c652e696f2f6c67746d2e706e67>
>
> @cuongdo <https://github.com/cuongdo> in line with the tests you're
> running right now, one that starts a single-node cluster with a wildly
> oversized range would be useful. This is easy to obtain by changing the
> zone config before creating the data dump, and then changing it back as you
> run the test.
>
> *Previously, d4l3k (Tristan Rice) wrote…*
>
> storage: Carve chunks of maxBytes / 2 off a range when splitting instead
> of splitting in the middle.
>
> This minimizes the amount of work done when splitting huge ranges as only
> ~32MB>
>
> needs to have stats recomputed instead of half of a potentially >1GB
> range.>
>
> Fixes #6400 <https://github.com/cockroachdb/cockroach/issues/6400>.
>
> ------------------------------
>
> Reviewed 3 of 3 files at r1.
> Review status: all files reviewed at latest revision, 1 unresolved
> discussion, some commit checks failed.
> ------------------------------
>
> *storage/engine/mvcc_test.go, line 2745 [r1]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7118#-KJloXGVxnY3ZPlHuTuL:-KJloXGWuq7ENPW_cDTq:-2041185795>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/baab0ac3cd5dd770f1bb43420be104e8be670487/storage/engine/mvcc_test.go#L2745>):*
>
>       snap := engine.NewSnapshot()
>       defer snap.Close()
>       targetSize := (ms.KeyBytes + ms.ValBytes) / 2
>
> Could you do a sanity check that a targetSize of 0 and \infinity behaves
> accordingly?
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7118#-:-KJlnxw6CePn774ulc2W:-720568726>*
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/7118#issuecomment-224727023>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABffppXYQw7XnXjJUR6Iszc6lglzA4xXks5qJy5BgaJpZM4IxZG2>
> .
>

tschottdorf commented:
https://github.com/cockroachdb/cockroach/issues/7121
-- 

-- Tobias
